File: elixir/pascals-triangle/pascals_triangle.exs
defmodule PascalsTriangle do
  @doc """
  Calculates the rows of a pascal triangle
  with the given height
  """
  @spec rows(integer) :: [[integer]]
  def rows(num), do: do_rows(num, 0, []) |> Enum.reverse

  defp do_rows(rows, rows, acc), do: acc
  defp do_rows(rows, cur, acc), do: do_rows( rows, cur+1, [ row(List.first(acc)) | acc ] )

  defp row(nil), do: [1]
  defp row(prev_row) do
    prev_row
    |> Enum.chunk(2, 1)
    |> Enum.map(fn [a,b] -> a + b end)
    |> List.insert_at(0, 1)
    |> List.insert_at(-1, 1)
  end
end

File: go/pascals-triangle/pascals_triangle.go
package pascal

func Triangle(rows int) [][]int {
	triangle := make([][]int, rows)

	for i := 0; i < rows; i++ {
		line := make([]int, i+1)

		for j := 0; j < i+1; j++ {
			if j == 0 || j == i {
				line[j] = 1
			} else {
				line[j] = triangle[i-1][j-1] + triangle[i-1][j]
			}
		}

		triangle[i] = line
	}

	return triangle
}
